tests/Microsoft.Identity.Web.Test/WebAppExtensionsTests.cs

```diff
@@ -1,32 +1,493 @@
+
+            Assert.Equal(_cookieScheme, oidcOptions.SignInScheme);
+            Assert.NotNull(oidcOptions.Authority);
+            Assert.NotNull(oidcOptions.TokenValidationParameters.IssuerValidator);
+            Assert.Equal(ClaimConstants.PreferredUserName, oidcOptions.TokenValidationParameters.NameClaimType);
+        }
+
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public async Task AddSignIn_WithConfigName_OnRedirectToIdentityProviderEvent(bool useServiceCollectionExtension)
+        {
+            var configMock = Substitute.For<IConfiguration>();
+            configMock.Configure().GetSection(_configSectionName).Returns(_configSection);
+
+            var redirectFunc = Substitute.For<Func<RedirectContext, Task>>();
+            var services = new ServiceCollection()
+                .Configure<OpenIdConnectOptions>(_oidcScheme, (options) => {
+                    options.Events ??= new OpenIdConnectEvents();
+                    options.Events.OnRedirectToIdentityProvider += redirectFunc;
+                });
+            services.AddDataProtection();
+
+            if (useServiceCollectionExtension)
+                services.AddSignIn(configMock, _configSectionName, _oidcScheme, _cookieScheme, false);
+            else
+                new AuthenticationBuilder(services)
+                    .AddSignIn(configMock, _configSectionName, _oidcScheme, _cookieScheme, false);
+
+            await AddSignIn_TestOnRedirectToIdentityProviderEvent(services, redirectFunc).ConfigureAwait(false);
+        }
+
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public async Task AddSignIn_WithConfigActions_OnRedirectToIdentityProviderEvent(bool useServiceCollectionExtension)
+        {
+            var redirectFunc = Substitute.For<Func<RedirectContext, Task>>();
+            var services = new ServiceCollection()
+                .Configure<OpenIdConnectOptions>(_oidcScheme, (options) => {
+                    options.Events ??= new OpenIdConnectEvents();
+                    options.Events.OnRedirectToIdentityProvider += redirectFunc;
+                });
+            services.AddDataProtection();
+
+            if (useServiceCollectionExtension)
+                services.AddSignIn(_configureOidcOptions, _configureMsOptions, _oidcScheme, _cookieScheme, false);
+            else
+                new AuthenticationBuilder(services)
+                    .AddSignIn(_configureOidcOptions, _configureMsOptions, _oidcScheme, _cookieScheme, false);
+
+            await AddSignIn_TestOnRedirectToIdentityProviderEvent(services, redirectFunc).ConfigureAwait(false);
+        }
+
+        private async Task AddSignIn_TestOnRedirectToIdentityProviderEvent(IServiceCollection services, Func<RedirectContext, Task> redirectFunc)
+        {
+            var provider = services.BuildServiceProvider();
```

Comment:
>var provider = services.BuildServiceProvider(); [](start = 12, length = 47)

in some tests you build the service provider in the test and pass it along, others you build it in the private method. can you move them all to the private method and reduce the amount of parameters?

Reply:
The reason why I build the provider in the public methods is because they have specific tests to that method that need a provider, like testing config by name vs by actions, and then I also need the provider in the private method to test the common things.